File: app/src/main/java/io/github/wulkanowy/utils/FabricUtils.java
package io.github.wulkanowy.utils;

import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.CustomEvent;
import com.crashlytics.android.answers.LoginEvent;
import com.crashlytics.android.answers.SignUpEvent;

public final class FabricUtils {

    private FabricUtils() {
        throw new IllegalStateException("Utility class");
    }

    public static void logLogin(String method, boolean result) {
        Answers.getInstance().logLogin(new LoginEvent()
                .putMethod(method)
                .putSuccess(result)
        );
    }

    public static void logRegister(boolean result, String symbol, String message) {
        Answers.getInstance().logSignUp(new SignUpEvent()
                .putMethod("Login activity")
                .putSuccess(result)
                .putCustomAttribute("symbol", symbol)
                .putCustomAttribute("message", message)
        );
    }

    public static void logRefresh(String name, boolean result, String date) {
        Answers.getInstance().logCustom(
                new CustomEvent(name + " refresh")
                        .putCustomAttribute("Success", result ? "true" : "false")
                        .putCustomAttribute("Date", date)
        );
    }
}

File: app/src/main/java/io/github/wulkanowy/utils/FabricUtils.kt
package io.github.wulkanowy.utils

import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.CustomEvent
import com.crashlytics.android.answers.LoginEvent
import com.crashlytics.android.answers.SignUpEvent
import timber.log.Timber
import kotlin.math.min

fun logLogin(method: String) {
    try {
        Answers.getInstance().logLogin(LoginEvent().putMethod(method))
    } catch (e: Throwable) {
        Timber.d(e)
    }
}

fun logRegister(message: String, result: Boolean, symbol: String, endpoint: String) {
    try {
        Answers.getInstance().logSignUp(SignUpEvent()
            .putMethod("Login activity")
            .putSuccess(result)
            .putCustomAttribute("symbol", symbol)
            .putCustomAttribute("message", message.substring(0, min(message.length, 100)))
            .putCustomAttribute("endpoint", endpoint)
        )
    } catch (e: Throwable) {
        Timber.d(e)
    }
}

fun <T> logEvent(name: String, params: Map<String, T>) {
    try {
        Answers.getInstance().logCustom(CustomEvent(name)
            .apply {
                params.forEach {
                    when {
                        it.value is String -> putCustomAttribute(it.key, it.value as String)
                        it.value is Number -> putCustomAttribute(it.key, it.value as Number)
                        it.value is Boolean -> putCustomAttribute(it.key, if ((it.value as Boolean)) "true" else "false")
                        else -> Timber.w("logEvent() unknown value type: ${it.value}")
                    }
                }
            }
        )
    } catch (e: Throwable) {
        Timber.d(e)
    }
}
